hoot-rnd/src/main/cpp/hoot/rnd/conflate/multiary/MultiaryClusterer.h

```diff
@@ -0,0 +1,180 @@
+     */
+    MatchClassification score;
+    /**
+     * Text that describes the link between the two clusters.
+     */
+    QString explainText;
+
+    ClusterLink(MultiaryClusterPtr a, MultiaryClusterPtr b, const MatchClassification& score)
+    {
+      this->a = a;
+      this->b = b;
+      this->score = score;
+    }
+
+    /**
+     * Returns true if both the clusters that are part of this link are valid.
+     */
+    bool isValid() const { return a->valid && b->valid; }
+
+    QString toString() const
+    {
+      return QString("{ 'score': %1, %2'a': %3, 'b': %4%5}").arg(score.toString()).
+        arg(isValid() ? "" : "'valid': false, ").
+        arg(a->toString(), b->toString()).
+        arg(explainText.isEmpty() ? "" : ", 'explainText': '" + explainText + "'");
+    }
+  };
+
+  typedef boost::shared_ptr<ClusterLink> ClusterLinkPtr;
+
+  /**
+   * A list of clusters that represent a subgraph.
+   */
+  typedef QList<MultiaryClusterPtr> ClusterList;
+
+  /**
+   * Comparison operation to determine if one link has a lower score than another. This is used
+   * when sorting and should always return a consistent ordering, even if the scores are identical.
+   */
+  struct ClusterLinkPtrLess
+  {
+    bool operator()(const ClusterLinkPtr& x, const ClusterLinkPtr& y) const
+    {
+      bool result;
+      if (x->score.getMatchP() == y->score.getMatchP())
+      {
+        if (x->score.getReviewP() < y->score.getReviewP())
+        {
+          result = true;
+        }
+        else if (x->score.getReviewP() > y->score.getReviewP())
+        {
+          result = false;
+        }
+        // in the case of an equal score, return a consistent less than value based on the element
+        // ID
+        else
+        {
+          result = x->a->at(0)->getElementId() < y->a->at(0)->getElementId();
+        }
+      }
+      else
+      {
+        result = x->score.getMatchP() < y->score.getMatchP();
+      }
+      return result;
+    }
+  };
+
+  /**
+   * @param mergeCache The cache to use when merging POIs. This abstracts the actual merge operation
+   *  away from the clustering algorithm.
+   * @param scoreCache The cache to use when scoring POIs. This abstracts the actual score operation
+   *  away from the clustering algorithm.
+   */
+  MultiaryClusterer(MultiaryPoiMergeCachePtr mergeCache, MultiaryScoreCachePtr scoreCache);
+
+  /**
+   * Cluster a subgraph of elements into a list of clusters. This will also populate the list of
+   * reviews which can be queried with takeReviews().
+   *
+   * Only direct pairs will be evaluated for cluster links so the list of pairs should accurately
+   * reflect the elements that have high scores.
+   *
+   * @param map The map that contains all the elements.
+   * @param pairs The pairs of elements that make up the subgraph.
+   * @return A list of clusters that were found in the 'pairs' subgraph.
+   */
+  virtual ClusterList cluster(OsmMapPtr map,
```

Comment:
First of all, know that I hate it when code reviewers gripe about names. So I am hating myself as I type this... but I worry about the use of the word 'cluster' as both a noun and a verb in this code. Maybe, to reduce cognitive load (Ha! Buzzwords!), consider using a different verb. Perhaps 'group' would be a good substitute in this instance.